server.go

```diff
@@ -104,11 +105,30 @@ func logResponse(status int, msg string) {
 }
 
 func respondWithImage(r *http.Request, rw http.ResponseWriter, data []byte, imgURL string, po processingOptions, duration time.Duration) {
-	gzipped := strings.Contains(r.Header.Get("Accept-Encoding"), "gzip") && conf.GZipCompression > 0
 
 	rw.Header().Set("Expires", time.Now().Add(time.Second*time.Duration(conf.TTL)).Format(http.TimeFormat))
 	rw.Header().Set("Cache-Control", fmt.Sprintf("max-age=%d, public", conf.TTL))
 	rw.Header().Set("Content-Type", mimes[po.format])
+
+	if conf.CacheFiles {
+
+		// calculate current ETag value using sha1 hashing function
+		currentEtagValue := fmt.Sprintf("%x", sha1.Sum(data))
+		rw.Header().Set("ETag", currentEtagValue)
+
+		// if client has its own locally cached copy of this file, then return 304, no need to send it again over the network
+		if currentEtagValue == r.Header.Get("If-None-Match") {
+			rw.WriteHeader(304)
+			logResponse(304, fmt.Sprintf("Returned cached image in %s: %s; %+v", duration, imgURL, po))
+			return
+		}
+
+	}
+
+	rw.Header().Set("Last-Modified", time.Now().Format(http.TimeFormat))
+
+	gzipped := strings.Contains(r.Header.Get("Accept-Encoding"), "gzip") && conf.GZipCompression > 0
```

Comment:
Please, don't move lines of code without need, even if you don't like their position.